### PR TITLE
feat: convert invite dialog to react and redux

### PR DESCRIPTION
**Pull request by virtuacoplenny**
The old invite modal has been removed and has been replaced with a version that uses react and redux. The behavior of the invite modal has changed a little, with the add/remove password components having a show/hide toggle.

Related jitsi-meet-torture changes: https://github.com/jitsi/jitsi-meet-torture/pull/79

### CODE REVIEW

File: react/features/base/conference/reducer.js [hunk 1]
```diff
@@ -209,7 +209,10 @@ function _lockStateChanged(state, action) {
         return state;
     }
 
-    return setStateProperty(state, 'locked', action.locked || undefined);
+    return setStateProperties(state, {
+        locked: action.locked || undefined,
+        roomLockPassword: action.locked ? state.roomLockPassword : null
```
Comment (yanas) [inline react/features/base/conference/reducer.js]: It bothers me a little bit that one is called "locked" and doesn't have "room" in the name and the other "roomLockPassword". Let's either rename locked to "roomLocked" or rename "roomLockPassword" to "lockPassword".

File: react/features/base/conference/middleware.js [hunk 2]
```diff
@@ -13,7 +18,12 @@ import {
     _setAudioOnlyVideoMuted,
     setLastN
 } from './actions';
-import { SET_AUDIO_ONLY, SET_LASTN } from './actionTypes';
+import {
+    LOCK_STATE_CHANGED,
+    SET_AUDIO_ONLY,
+    SET_LASTN,
+    SET_PASSWORD_FAILED
```
Comment (yanas) [inline react/features/base/conference/middleware.js]: Shouldn't LOCK_STATE_CHANGED AND SET_PASSWORD_FAILED be part of the lock-room middleware? @lyubomir should we try to keep those within the feature?

File: lang/main.json [hunk 3]
```diff
@@ -428,5 +428,13 @@
         "noOtherDevices": "No other devices available",
         "selectADevice": "Select a device",
         "testAudio": "Test sound"
+    },
+    "invite": {
+        "addPassword": "Add password",
+        "hidePassword": "Hide password",
```
Comment (yanas) [inline lang/main.json]: Can we make sure that strings that are not used anymore are deleted. I see:
"addPassword": "Add a password",
"createPassword": "Create password",
in the same file.

File: react/features/invite/components/LockState.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import React, { Component } from 'react';
+
+import { translate } from '../../base/i18n';
+
+/**
+ * A React Component for displaying the conference lock state.
+ */
+class LockState extends Component {
```
Comment (yanas) [inline react/features/invite/components/LockState.js]: I realise that we don't have a naming convention on Components or maybe we have and I don't know about it, but I would suggest that we add some word at the end that indicates what kind of component is rendered in the file. For example AddPassword.js will become AddPasswordContainer or AddPasswordForm and RemovePassword.js will become RemovePasswordContainer. PasswordOverview.js will become PasswordContainer probably and LockState will be LockStatePanel or LockStateLabel.

File: lang/main.json (hunk 3 above)
Comment (virtuacoplenny) [inline lang/main.json]: I took a pass just took another pass and removed a couple strings related to room locking that weren't being used. createPassword is being used as the placeholder in the input for adding a password.

File: react/features/base/dialog/components/Dialog.web.js [hunk 5]
```diff
@@ -56,6 +56,7 @@ class Dialog extends AbstractDialog {
                 width = { this.props.width || 'medium' }>
                 <div>
                     <form
+                        className = 'modal-dialog-form'
                         id = 'modal-dialog-form'
```
Comment (lyubomir) [inline react/features/base/dialog/components/Dialog.web.js]: Given that the `id` is required (I guess by `<AKButton form = 'modal-dialog-form'>`) why is a className necessary?

File: react/features/base/dialog/components/AbstractDialog.js [hunk 6]
```diff
@@ -56,7 +56,12 @@ export default class AbstractDialog extends Component {
         /**
          * Key to use for showing a title.
          */
-        titleKey: React.PropTypes.string
+        titleKey: React.PropTypes.string,
+
+        /**
+         * String to use as a title instead of a titleKey.
```
Comment (lyubomir) [inline react/features/base/dialog/components/AbstractDialog.js]: Please expand the comment to clarify that specifying a truthy `titleString` value will take precedence over `titleKey`.

File: react/features/base/conference/reducer.js [hunk 7]
```diff
@@ -209,7 +209,10 @@ function _lockStateChanged(state, action) {
         return state;
     }
 
-    return setStateProperty(state, 'locked', action.locked || undefined);
+    return setStateProperties(state, {
+        locked: action.locked || undefined,
+        lockPassword: action.locked ? state.lockPassword : null
```
Comment (lyubomir) [inline react/features/base/conference/reducer.js]: @virtuacoplenny, I'm confused by the introduction of `lockPassword` in the (redux) state of the conference alongside the existing `password`.

1. Generally, there's one password in a conference in the context of `conference.join` & `conference.lock` - the password one uses with `conference.lock` is what others will have to use with `conference.join`. If `conference.join` or `conference.lock` complete successfully, the password they completed successfully with should become `password`.

2. Depending on the implementation, I can image one may need a "temporary password" from between the time the user specifies a password to be utilized for `conference.lock` and the time of the successful completion of `conference.lock`. But once `conference.lock` completes successfully, that "temporary password" should surely turn into `password`. I don't expect such a "temporary password" to necessarily have to be a (redux) state of the password, I'd rather expect it to be a state of some sort of some other, probably UI-related feature. Anyway, (1) I don't think that's your idea of `lockPassword` and (2) even if that was your idea, at the every least some non-trivial jsdoc comments would've been necessary.

Could you please remove `lockPassword` in order to bring clarity?

**Comment by virtuacoplenny** [comment]
@lyubomir, I left the changes of having only one password as a separate commit so you could review it separately. I moved it before your style changes so it can be squashed.

File: react/features/base/conference/reducer.js [hunk 8]
```diff
@@ -209,7 +209,11 @@ function _lockStateChanged(state, action) {
         return state;
     }
 
-    return setStateProperty(state, 'locked', action.locked || undefined);
+    return setStateProperties(state, {
+        locked: action.locked || undefined,
+        lockedLocally: Boolean(state.lockedLocally && action.locked),
```
Comment (lyubomir) [inline react/features/base/conference/reducer.js]: I'm bothered by the lack of symmetry in having `lockedLocally` and not having `lockedRemotely`. I think I'd have been more at peace with 1 single `locked` state which evaluates to an enumerable set of values such as:
- If `locked` is falsy, then the conference is unlocked.
- If `locked` is truthy, then the conference is locked. And if I as the API consumer care about how it's locked, I can go beyond truthy by examining more specifics of the value such as is it equal to LOCKED_LOCALLY or LOCKED_REMOTELY.
WDYT, @virtuacoplenny?

File: react/features/invite/components/ShareLinkForm.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * Initializes a new ShareLinkForm instance.
+     *
+     * @param {Object} props - The read-only properties with which the new
+     * instance is to be initialized.
+     */
+    constructor(props) {
+        super(props);
+
+        this._inputElement = null;
```
Comment (lyubomir) [inline react/features/invite/components/ShareLinkForm.js]: Technically, the above introduces a field and our coding style requires jsdocs on everything which includes fields. Anyway, I'll write this one in order to spare time.

File: react/features/invite/components/PasswordContainer.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+        /**
+         * Whether or not the conference is currently locked.
+         */
+        locked: React.PropTypes.bool,
+
+        /**
+         * Whether or not the room lock, if any, was set by the local user.
+         */
+        lockedLocally: React.PropTypes.bool,
```
Comment (lyubomir) [inline react/features/invite/components/PasswordContainer.js]: Ugh, you have to set both `locked` and `lockedLocally`. If you set `lockedLocally` to true, the code will still treat the conference as unlocked. I feel like my idea about an enum fits here so much better.

File: react/features/invite/components/PasswordContainer.js [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+     * Initializes a new PasswordContainer instance.
+     *
+     * @param {Object} props - The read-only properties with which the new
+     * instance is to be initialized.
+     */
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            isEditingPassword: false
```
Comment (lyubomir) [inline react/features/invite/components/PasswordContainer.js]: Technically, that's a field and our coding style requires a jsdoc on it.

File: react/features/invite/components/LockStatePanel.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    /**
+     * Implements React's {@link Component#render()}.
+     *
+     * @inheritdoc
+     * @returns {ReactElement}
+     */
+    render() {
+        const [ lockStateClass, lockIconClass, lockTextKey ] = this.props.locked
+            ? [ 'is-locked', 'icon-security-locked', 'invite.locked' ]
+            : [ 'is-unlocked', 'icon-security', 'invite.unlocked' ];
```
Comment (lyubomir) [inline react/features/invite/components/LockStatePanel.js]: May I please ask you to change that to an `if/else` with simple variable assignments?

File: react/features/invite/components/InviteDialog.js [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+ * @returns {{
+ *     _conference: Object,
+ *     _localParticipant: Object
+ * }}
+ */
+function _mapStateToProps(state) {
+    return {
+        _conference: state['features/base/conference'],
+        _localParticipant:
+            getLocalParticipant(state['features/base/participants'])
```
Comment (lyubomir) [inline react/features/invite/components/InviteDialog.js]: I'm not sure why you're mapping the whole feature states into props.  I guess you could see how it makes sense for `_conference` (or maybe not), I don't see any sense in `_localParticipant`.

**Comment by virtuacoplenny** [comment]
@lyubomir, keeping remote vs local within locked resulted in a lot of good cleanup. Thanks for the suggestion. I moved that change, as well as a change to not pass in the full participant, before your styling changes. My changes should be squashed, which I can do or you can if you want. I also made some of the styling changes you requested and a few others and put that after your changes, and I think it should be squashed into yours if you agree with them.

**Pull request merged**